libyul/backends/wasm/EVMToEwasmTranslator.cpp

```diff
@@ -806,9 +806,346 @@ function u256_to_address(x1, x2, x3, x4) -> r1, r2, r3 {
 	r3 := x4
 }
 
+function i8_store(ptr:i32, v:i64)
+{
+	let tmp := i64.and(i64.load(ptr), 0x00ffffffffffffff)
+	tmp := i64.or(tmp, i64.shl(i64.and(v, 0xff), 56))
+	i64.store(ptr, endian_swap(tmp))
+}
+
+function i8_load32(ptr:i32) -> v:i32
+{
+	v := i32.and(i32.load(ptr), 0x000000ff:i32)
+}
+
+function i8_load(ptr:i32) -> v:i64
+{
+	v := i64.and(i64.load(ptr), 0x00000000000000ff)
+}
+
 function keccak256(x1, x2, x3, x4, y1, y2, y3, y4) -> z1, z2, z3, z4 {
-	// TODO implement
-	unreachable()
+	let input:i32 := u256_to_i32ptr(x1, x2, x3, x4) // 0xa0
+	let len:i64 := u256_to_i64(y1, y2, y3, y4)      // 0x00
+
+	// 0) Setup Constants & Data
+	// -------------------------
+	//    const uint64_t keccakf_rndc[24] = {
+	//        0x0000000000000001, 0x0000000000008082, 0x800000000000808a,
+	//        0x8000000080008000, 0x000000000000808b, 0x0000000080000001,
+	//        0x8000000080008081, 0x8000000000008009, 0x000000000000008a,
+	//        0x0000000000000088, 0x0000000080008009, 0x000000008000000a,
+	//        0x000000008000808b, 0x800000000000008b, 0x8000000000008089,
+	//        0x8000000000008003, 0x8000000000008002, 0x8000000000000080,
+	//        0x000000000000800a, 0x800000008000000a, 0x8000000080008081,
+	//        0x8000000000008080, 0x0000000080000001, 0x8000000080008008
+	//    };
+	//
+	//////////////////////
+
+	mstore_internal(0x40:i32, 0xF000, 0x0000, 0x0000, 0x0000) // free-memory pointer
+
+	/////////////////////
+	//
+	let m1, m2, m3, m4 := mload_internal(0x40:i32)
+	let slot:i32 := u256_to_i32ptr(m4, m3, m2, m1) // from this position on we will need 512 bytes
+	mstore_internal_raw(i32.add(slot, 0x00:i32), 0x0000000000000001, 0x0000000000008082, 0x800000000000808a, 0x8000000080008000)
```

Comment:
Why don't you just rely on `i64.store` to store the array?

Reply:
I general I'd try to keep this code closer to wasm and less using the evm->wasm helpers. Should improve understandability, given wasm opcodes 32/64-bit, while evm is 256-bit.